site/src/lib/api/blocks/npm.ts

```diff
@@ -0,0 +1,278 @@
+      cwd: npmTarballFolder,
+      file: path.resolve(npmTarballFolder, tarballFilename),
+    });
+  } catch (err) {
+    throw new Error(
+      `Could not find package '${npmPackageName}'. Does it exist?`,
+      { cause: { code: "PACKAGE_NOT_FOUND" } },
+    );
+  }
+
+  // check the package contents and retrieve useful metadata
+  const packageFolder = path.resolve(npmTarballFolder, "package");
+
+  // get the package.json - we know this exists and is valid JSON, since it's an npm-published package
+  const packageJsonString = fs
+    .readFileSync(path.resolve(packageFolder, "package.json"))
+    .toString();
+  const packageJson = JSON.parse(packageJsonString);
+
+  // get the block-metadata.json
+  const metadataJsonPath = glob.sync("**/block-metadata.json", {
+    cwd: packageFolder,
+    nocase: true,
+  })[0];
+  if (!metadataJsonPath) {
+    throw new Error("No block-metadata.json present in package", {
+      cause: { code: "INVALID_PACKAGE_CONTENTS" },
+    });
+  }
+
+  let metadataJson;
+  try {
+    const metadataJsonString = fs
+      .readFileSync(path.resolve(packageFolder, metadataJsonPath))
+      .toString();
+    metadataJson = JSON.parse(metadataJsonString);
+  } catch (err) {
+    throw new Error(
+      `Could not parse block-metadata.json: ${(err as Error).message}`,
+      {
+        cause: { code: "INVALID_PACKAGE_CONTENTS" },
+      },
+    );
+  }
+
+  const blockSourceFolder = path.resolve(
+    packageFolder,
+    path.dirname(metadataJsonPath),
+  );
+
+  // move the readme into the block's source folder, if it exists
+  const readmePath = glob.sync("README.md", {
+    cwd: packageFolder,
+    nocase: true,
+  })[0];
+  if (readmePath) {
+    fs.renameSync(
+      path.resolve(packageFolder, readmePath),
+      path.resolve(blockSourceFolder, path.basename(readmePath)),
+    );
+  }
+
+  // check if we have 'example-graph.json', which we then construct a URL for in the extended matadata
+  const includesExampleGraph = await fs.pathExists(
+    path.resolve(blockSourceFolder, "example-graph.json"),
+  );
+
+  // check block-metadata.json contains required properties
+  for (const key of ["blockType", "protocol", "schema", "source", "version"]) {
+    if (!metadataJson[key]) {
+      throw new Error(`block-metadata.json must contain a '${key}' property`, {
+        cause: { code: "INVALID_PACKAGE_CONTENTS" },
+      });
+    }
+  }
+
+  // check that the source file actually exists
+  const sourcePath = metadataJson.source;
+  const sourceFileExists = await fs.pathExists(
+    path.resolve(blockSourceFolder, sourcePath),
+  );
+
+  if (!sourceFileExists) {
+    throw new Error(
+      `block-metadata.json 'source' path '${sourcePath}' does not exist`,
+      {
+        cause: { code: "INVALID_PACKAGE_CONTENTS" },
+      },
+    );
+  }
+
+  // check that the schema actually exists
+  const schemaPath = metadataJson.schema;
+  if (schemaPath.startsWith("http")) {
+    await fetch(schemaPath, { method: "HEAD" });
```

Comment:
`@todo` – this error should be caught and a friendly error message/code provided